### PR TITLE
Add tests for JSX

### PR DESCRIPTION
**Pull request by ulrikstrid**
These tests are copied and fixed up from the ppx tests.
I removed the whitespace tests because it doesn't make sense since JSX fixes that out of the box.
I also commented out the comment tests since I'm not sure how to fix them.

### CODE REVIEW

File: test/test_jsx.re [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+        [<> <p /> <span> foo </span> </>],
+        [p([]), span([txt("foo")])],
+      ),
+      ("attrib", [[<p id=foo />]], [p(~a=[a_id("foo")], [])]),
+      (
+        "attribs",
+        [[<p id=foo className=bar />]],
+        [p(~a=[a_id("foo"), a_class(["bar"])], [])],
+      ),
+      /* can this be done in JSX?
```
Comment (Khady) [inline test/test_jsx.re:78]: hmm probably something must be added to the ppx to support this

File: test/test_jsx.re [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+  let pp_wrap = pp => pp;
+  let pp_wrap_list = pp => Format.pp_print_list(~pp_sep=(_, ()) => (), pp);
+};
+module Svg = {
+  include Tyxml.Svg;
+  let pp_wrap = pp => pp;
+  let pp_wrap_list = pp => Format.pp_print_list(~pp_sep=(_, ()) => (), pp);
+};
+module HtmlTests = TyTests(Html);
+module SvgTests = TyTests(Svg);
```
Comment (Khady) [inline test/test_jsx.re:55]: probably all this setup could be shared with the ppx tests. But I don't know if it worth the trouble

File: test/test_jsx.re [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+          and y = <a />;
+          [x, y];
+        },
+        [p([]), a([])],
+      ),
+      (
+        "let fun",
+        [
+          {
+            let%html f = x => "<p>"(x, "</p>");
```
Comment (Khady) [inline test/test_jsx.re]: this one was not translated?

Comment (ulrikstrid) [reply test/test_jsx.re]: That's fixed, had 1 more that I missed as well

File: test/test_jsx.re [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+          ),
+          ("first child", [(elt1())(<p />)], elt1() @- p(nil()) @: nil()),
+          ("last child", [<> <p /> {elt1()} </>], p(nil()) @: elt1()),
+          (
+            "wrapped functions",
+            !:[<input method=`Get />],
+            !:input(~a=[a_method(`Get^)], ()),
+          ),
+        ],
+        /* should succeed */
```
Comment (Khady) [inline test/test_jsx.re]: what's wrong with this one?

Comment (ulrikstrid) [reply test/test_jsx.re]: It was commented out preciously, I can just add it in I guess

File: test/test_jsx.re [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+        "with_neighbour",
+        [[<div> <span /> <svg> <g /> </svg> foo </div>]],
+        [div([span([]), svg([Svg.g([])]), txt("foo")])],
+      ),
+      ("ambiguous tag", [[<svg> <a /> </svg>]], [svg([Svg.a([])])]),
+    ],
+  ),
+);
+
+let svg = (
```
Comment (Khady) [inline test/test_jsx.re:272]: those ones are not translated?

File: test/test_jsx.re [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+        (
+          "wrapped functions",
+          !:[<input method=`Get />],
+          !:input(~a=[a_method(`Get^)], ()),
+        ),
+      ],
+      /* should succeed */
+      /* "escape", */
+      /* [%tyxml "<p>(tyxml4)</p>"], */
+      /* [p [txt "(tyxml4)"]]; */
```
Comment (Drup) [inline test/test_jsx.re]: You can remove this, it doesn't make sense for the jsx :)

File: test/test_jsx.re (hunk 2 above)
Comment (Drup) [reply test/test_jsx.re:55]: Well, it's copy-pasted from the ppx tests. @ulrikstrid Is there a reason not to import from the ppx tests and/or factor in a test library ?

File: test/test_jsx.re [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+module SvgTests = TyTests(Svg);
+
+let basics = (
+  "ppx basics",
+  HtmlTests.make(
+    Html.[
+      ("elems", [[<p />]], [p([])]),
+      (
+        "child",
+        [[<p> <span> foo </span> </p>]],
```
Comment (Drup) [inline test/test_jsx.re:64]: Why are those list of lists ? Am I misunderstanding the syntax ?

File: ppx/jsx-ppx/tyxml_jsx.ml [hunk 8]
```diff
@@ -16,6 +17,58 @@ let make_html_attr_name name =
...
+    | "ariaDetails" -> "aria-details"
+    | "ariaDisabled" -> "aria-disabled"
+    | "ariaHidden" -> "aria-hidden"
+    | "ariaInvalid" -> "aria-invalid"
+    | "ariaKeyshortcuts" -> "aria-keyshortcuts"
+    | "ariaLabel" -> "aria-label"
+    | "ariaRoledescription" -> "aria-roledescription"
+    (* Widget Attributes *)
+    | "ariaAutocomplete" -> "aria-autocomplete"
+    | "ariaChecked: true|false|mixed, /* https" -> "aria-checked"
```
Comment (Khady) [inline ppx/jsx-ppx/tyxml_jsx.ml]: hmm I'm not sure this one will work, should simply be `"ariaChecked"`?

File: ppx/jsx-ppx/tyxml_jsx.ml [hunk 9]
```diff
@@ -16,6 +17,58 @@ let make_html_attr_name name =
...
+    | "ariaExpanded" -> "aria-expanded"
+    | "ariaHaspopup" -> "aria-haspopup"
+    | "ariaLevel" -> "aria-level"
+    | "ariaModal" -> "aria-modal"
+    | "ariaMultiline" -> "aria-multiline"
+    | "ariaMultiselectable" -> "aria-multiselectable"
+    | "ariaOrientation" -> "aria-orientation"
+    | "ariaPlaceholder" -> "aria-placeholder"
+    | "ariaPressed: true|false|mixed, /* https" -> "aria-pressed"
```
Comment (Khady) [inline ppx/jsx-ppx/tyxml_jsx.ml]: some as above

File: test/test_jsx.re [hunk 10]
```diff
@@ -0,0 +1,562 @@
...
+      ),
+      /* can this be done in JSX?
+         ("comment", [[%html "<!--foo-->"]], [tot @@ Xml.comment("foo")]),
+         */
+      ("txt", [[foo]], [txt("foo")]),
+      (
+        "document",
+        [[<html> <head> <title> foo </title> </head> <body /> </html>]],
+        [html(head(title(txt("foo")), []), body([]))],
+      ),
```
Comment (Drup) [inline test/test_jsx.re:86]: Could you add a test for nullary, unary and star elements ?

**Review by Drup — COMMENTED** [review]
Looks pretty good. I think you should thin it down a little bit actually, remove the things that are element-specific and will be exercised by the other tests. I want the test here to be only for mostly-syntax, since most of the implementation after the initial "parsing" are common.

Otherwise, I made some notes which you can fix either now or later.

File: ppx/jsx-ppx/tyxml_jsx.ml [hunk 11]
```diff
@@ -5,6 +5,7 @@ open Asttypes
 open Tyxml_ppx_common
 
 let make_html_attr_name name =
+  (* Aria attributes are the same as reason-react https://github.com/reasonml/reason-react/blob/5ce8498207ca1f2e9b8305cfca231c837b2d021b/src/ReactDOMRe.re *)
```
Comment (Drup) [inline ppx/jsx-ppx/tyxml_jsx.ml:8]: This treatement of aria is not good, and not in line with what we do in tyxml anyway. Instead, use a general transform that turns everything of the form `aria*` into `aria-lowercase(*)`.

File: ppx/jsx-ppx/tyxml_jsx.ml (hunk 8 above)
Comment (ulrikstrid) [reply ppx/jsx-ppx/tyxml_jsx.ml]: I was a bit too fast with my find and replace...

**Comment by ulrikstrid** [comment]
Closing since it moved to the tyxml repo

**Pull request closed without merging**